src/cc_catalog_airflow/dags/util/popularity/sql.py

```diff
@@ -0,0 +1,69 @@
+from textwrap import dedent
+from airflow.hooks.postgres_hook import PostgresHook
+
+
+def upload_normalized_popularity(postgres_conn_id, in_tsv):
+    """
+    Write the `normalized_popularity` field from `in_tsv` to the catalog.
+    """
+    postgres = PostgresHook(postgres_conn_id=postgres_conn_id)
+    postgres.copy_expert(
+        "CREATE TEMP TABLE temp_popularity "
+        "  (identifier uuid, normalized_popularity text); "
+        "COPY temp_popularity FROM STDIN WITH CSV HEADER DELIMITER E'\t'; "
+        "UPDATE image SET meta_data = jsonb_set("
+        "  meta_data, "
+        "  '{normalized_popularity}',"
+        "  temp_popularity.normalized_popularity::jsonb"
+        ") FROM temp_popularity"
+        "  WHERE image.identifier = temp_popularity.identifier;",
+        in_tsv
+    )
+
+
+def select_percentiles(postgres_conn_id, popularity_fields, percentile):
+    """
+    Given a list of fields that occur in the `meta_data` column, return a dict
+    mapping each field to its `percentile`th percentile value.
```

Comment:
As I'm thinking about it, there are four apparent options for the set over which the percentile value should be calculated:
1. Globally:  This would treat `views` the same, no matter what site they're on.  So, views on Flickr would count the same as views on Behance.
   - Good:  
     - Fair, easy to track data.
     - If we collect data from multiple providers for the same image, we'd be able to combine such metrics simply by summing.
     - easy interpretation. 
   - Bad:
     - We lose the 'pinning a percentile to a value' effect on the individual provider and source distributions.
     - We are no longer in control of the shape of the distributions between different sources (providers).
     - In the case we, say, collect views of one NASA image from Flickr, and views of another NASA image from nasa.gov, the output popularity metric comparison between those two images will have more to do with the relative popularity of flickr.com vs nasa.gov than it will with the popularity of those images within those respective sites.  If we consider both to be from the source `nasa`, then weighting that source won't be able to mitigate that effect (although weighting providers would).
2. Per-provider:  We treat all `views` the same if they're collected from a given provider.
   - Good:
     - This removes the last 'bad' point above, since the 'best' image from flickr.com will get approximately the same popularity score as the 'best' image from nasa.gov
     - We regain the 'pinning a percentile to a value' effect for the distribution within each provider.
     - We 'standardize' the distributions between providers (Flickr vs Behance).
   - Bad:
     - Harder to interpret output.
     - We add complexity.
     - Still no 'pinning a percentile to a value' for a source distribution
     - The 'best' NASA image from flickr.com will have a different score than the best NASA image from nasa.gov
     - No control of shape of distribution between different sources (NASA vs SI)
3. Per-source:  We treat all `views` the same if they're images posted somewhere by the same source.
   - Good:
     - We get the 'pinning a percentile to a value' effect for images posted by a given source.
     - We standardize the distributions between sources.
   - Bad:
     - Add complexity
     - No 'pinning a percentile to a value' within a provider, leading to the next bullet point problem.
     - Given we collect images from more than one provider for a given source, the ranking of images internal to that provider will have more to do with the relative popularity of the providers involved than their ranking within those providers (same problem as option (1) above).
4. Per-(source, provider) pair:  We treat all `views` the same if they're of images for a given source, on a given provider site.
   - Good:
     - Granular control
     - Removes the last 'bad' point of options (1) and (3).
     - We standardize distributions between (source, provider) pairs:  This means we'd have the 'best' NASA image from Flickr next to the 'best' NASA image from nasa.gov.
   - Bad:
     - Hard to interpret
     - Complexity
     - No pinning a percentile to a value within a provider or within a source (unless the source is the provider for all images from that source, and that provider)
   - Not sure if good or bad:
     - This would have the effect of naturally 'bumping up' the score of images posted by a given source on lower-traffic provider sites.

Reply:
@annatuma This is feeling more-and-more like a product question.  What do you think?

Reply:
Currently I'm leaning towards option 4, as it strikes me as the most "fair" for what we want to do. If not 4 (e.g. due to complexity), then 2.

Definitely not 1 (global).

However, we'll discuss later today in more detail.

Reply:
Agreed that (1) is not an option. I suggest we go with (2) until we are able to observe the actual impact that normalized popularity has on search rankings. Since we only have views from Flickr (Behance is disabled), we don't have to do anything to enforce that. In the future, when you incorporate a new metric, you can call it `flickr_views`, `behance_views`, etc. . . assume a global namespace independent of sources.

Reply:
@aldenstpage I'd rather not need to handle the difference at the data level (e.g., with `flickr_views`, `behance_views`, etc.).  The reason is that this would require the implementer of any script that wanted to collect such a metric to have more global knowledge than should be necessary.  They'd need to know each of the other tags in order to ensure there was no conflict.  Also, whoever wanted to incorporate this into the normalized popularity metric would need to know all of the different names, rather than just partitioning by source or provider (or whatever) and tallying some 'standard' column names like views, comments, etc.

However, I do think that we can try it in the current config to see (if nothing else) how long of a calculation we're looking at, and to get some results in the front end so we can see the improvement.